dashboard home

gumaerc opened the pull request:
### What are the relevant tickets?
Closes https://github.com/mitodl/hq/issues/4524

### Description (What does it do?)
This PR adds a number of `ResourceCarousel`s to the User Dashboard Home page:

 - A "top picks for you" carousel based on all of the search preferences of the current user's `Profile`: topics, certification offering and learning format
 - One for each topic in the user's `Profile`
 - A carousel for courses with / without certificates based on what the user selected in their `Profile`
 - Carousels for new and popular learning resources

### Screenshots (if appropriate):
![image](https://github.com/mitodl/mit-open/assets/12089658/3758d64f-04c1-425d-a7e2-9bceeac39eda)
![image](https://github.com/mitodl/mit-open/assets/12089658/c8dc8f3f-a1bb-42e9-9fc5-eb241dd7e89b)

### How can this be tested?
 - Spin up `mit-open` on this branch
 - Make sure you have some courses and other `LearningResource` objects in your local database by using any of the management commands for backpopulating data from the various providers
 - Make sure you are logged in and that your user has a `Profile` associated with it; this can be done either by following the onboarding process (http://localhost:8063/onboarding) or through Django Admin by creating it manually and assigning it to your user, making sure you provide some information in each section
 - Visit the site at http://localhost:8063/
 - Click your username in the header, then click "Dashboard"
 - Verify that you see all of the carousels described above, and that the courses / resources displayed within make sense in each category
 - Ensure the layout looks correct on both desktop and mobile

### Additional Context
In the design you may see some tabs above the carousels in the design. This has been deemed out of scope for now and will be addressed in the future.

mbilalmughal commented:
@gumaerc Small spacing & typo things I've noticed

space between title & cards on mobile - 16px
![image](https://github.com/mitodl/mit-open/assets/7486671/d884e9e0-5825-4208-b28c-d9320dd30d3d)

Font size for title on mobile - 18px
![image](https://github.com/mitodl/mit-open/assets/7486671/4bd0ee20-0e64-4331-9d02-59cdbb376c46)

![image](https://github.com/mitodl/mit-open/assets/7486671/2fcc3afa-05ca-4df4-bb74-7746f065efc1)


Space between the next sections - 32px
![image](https://github.com/mitodl/mit-open/assets/7486671/a399a2e3-5771-4563-a1c4-e5c83a17fdab)


spaces on top title
![image](https://github.com/mitodl/mit-open/assets/7486671/1d21b91a-6b8f-42e0-a24e-4b4d641d6daa)

I can't see the desktop properly for measurement, but it looks good to me.

gumaerc commented:
@mbilalmughal Thanks, I adjusted the spacing based on what you said above with the following caveats:

 - Carousel card and title spacing
 - Font size of carousel headings specifically

I wanted to hold off on this because I'm using the `ResourceCarousel` component that @ChristopherChudzicki made, and changing styles on the component will change them everywhere including the home page. I just wanted to make sure we were okay with that before I went ahead and changed it.

I looked at the page header ("Your MIT Learning Journey") and the font size is indeed set to `h5` on mobile:

```
const TitleText = styled(Typography)(({ theme }) => ({
  color: theme.custom.colors.black,
  ...theme.typography.h3,
  [theme.breakpoints.down("md")]: {
    ...theme.typography.h5,
  },
}))

const SubTitleText = styled(Typography)(({ theme }) => ({
  color: theme.custom.colors.darkGray2,
  ...theme.typography.body1,
  [theme.breakpoints.down("md")]: {
    ...theme.typography.subtitle3,
  },
}))
```

So, if you're seeing a discrepancy than likely there is a problem with the fonts themselves and how they appear in Figma vs reality on the web.

mbilalmughal commented:
@gumaerc @shanbady After reviewing https://open.c4103.com/dashboard#home, everything (Desktop & Mobile both) looks good to me.

ChristopherChudzicki commented on frontends/mit-open/src/pages/DashboardPage/Dashboard.test.tsx:
```diff
@@ -74,20 +220,89 @@ describe("DashboardPage", () => {
+    const newCarousel = await screen.findByTestId(
+      "new-learning-resources-carousel",
+    )
+    all.forEach(async (learningResource) => {
```
I suggest changing this to 
```
await Promise.all(all.map(...))
```
I'm not entirely sure it's guaranteed that these promises resolve before the test finishes.


That said, I'm not sure it's worth making very detailed assertions about the carousel like this. The resource carousel has its own tests. I generally do support the philosophy of "testing what the user should see", but for something we (1) test separately and (2) reuse so much, it seems appropriate to me to do

```ts
jest.mocked("@/page-components/ResourceCarousel/ResourceCarousel", () => {
  const actual = jest.requireActual(
    "@/page-components/ResourceCarousel/ResourceCarousel",
  )
  return {
    ...actual,
    __esModule: true,
    default: jest.fn(actual.default),
  }
})
const spyCarousel = jest.mocked(ResourceCarousel)
```
and asserting that the carousel has been passed appropriate configs based on user profile.

**OR, even simpler** Mock the carousel, don't just spy on it:
```
jest.mocked("@/page-components/ResourceCarousel/ResourceCarousel", () => {
  const actual = jest.requireActual(
    "@/page-components/ResourceCarousel/ResourceCarousel",
  )
  return {
    ...actual,
    __esModule: true,
    default: jest.fn(() => (<div />)), // I think you need to return a JSX element or react gets mad. I forget.
  }
})
```

The carousel is something we use so much, I think this is probably the way we should handle testing it on individual pages.

ChristopherChudzicki reviewed (approved):
This looks good to me. I did request a change around some really weird formatting in the test file. 

Additionally, I have noticed the following odd behavior that I think we should change but not necessarily here.

1. User navigate sto search page
2. Carousels load without a profile
3. Profile loads
4. Carousels reload with a profile

In practice I'm not sure (2) is really visible to the user—depends the order that API request are made. But it's easy to see this behavior in the network tab.

I do think it would be good to add a `loading` prop to the resource carousel so that you could prevent it from making the requests until the profile is loaded. We don't need to do that in this PR though.
